src/picongpu/include/plugins/IsaacPlugin.hpp

```diff
@@ -0,0 +1,437 @@
+        static const bool persistent = bmpl::not_<boost::is_same<FieldType, FieldJ > >::value;
+        typename FieldType::DataBoxType shifted;
+        MappingDesc *cellDescription;
+        bool movingWindow;
+        TFieldSource() : cellDescription(NULL), movingWindow(false) {}
+
+        void init(MappingDesc *cellDescription, bool movingWindow)
+        {
+            this->cellDescription = cellDescription;
+            this->movingWindow = movingWindow;
+        }
+
+        static std::string getName()
+        {
+            return FieldType::getName() + std::string(" field");
+        }
+
+        void update(bool enabled, void* pointer)
+        {
+            const SubGrid<simDim>& subGrid = Environment< simDim >::get().SubGrid();
+            DataConnector &dc = Environment< simDim >::get().DataConnector();
+            FieldType * pField = &(dc.getData< FieldType > (FieldType::getName(), true));
+            DataSpace< simDim > guarding = SuperCellSize::toRT() * cellDescription->getGuardingSuperCells();
+            if (movingWindow)
+            {
+                GridController<simDim> &gc = Environment<simDim>::get().GridController();
+                if (gc.getPosition()[1] == 0) //first gpu
+                {
+					uint32_t* currentStep = (uint32_t*)pointer;
+                    Window window( MovingWindow::getInstance().getWindow( *currentStep ) );
+                    guarding += subGrid.getLocalDomain().size - window.localDimensions.size;
+                }
+            }
+            typename FieldType::DataBoxType dataBox = pField->getDeviceDataBox();
+            shifted = dataBox.shift( guarding );
+        }
+
+        ISAAC_NO_HOST_DEVICE_WARNING
+        ISAAC_HOST_DEVICE_INLINE isaac_float_dim< feature_dim > operator[] (const isaac_int3& nIndex) const
+        {
+            auto value = shifted[nIndex.z][nIndex.y][nIndex.x];
+            isaac_float_dim< feature_dim > result =
+            {
+                isaac_float( value.x() ),
+                isaac_float( value.y() ),
+                isaac_float( value.z() )
+            };
+            return result;
+        }
+};
+
+ISAAC_NO_HOST_DEVICE_WARNING
+template< typename FrameSolver, typename ParticleType >
+class TFieldSource< FieldTmpOperation< FrameSolver, ParticleType > >
+{
+    public:
+        static const size_t feature_dim = 1;
+        static const bool has_guard = false;
+        static const bool persistent = false;
+        typename FieldTmp::DataBoxType shifted;
+        MappingDesc *cellDescription;
+        bool movingWindow;
+
+        TFieldSource() : cellDescription(NULL), movingWindow(false) {}
+
+        void init(MappingDesc *cellDescription, bool movingWindow)
+        {
+            this->cellDescription = cellDescription;
+            this->movingWindow = movingWindow;
+        }
+
+        static std::string getName()
+        {
+            return ParticleType::FrameType::getName() + std::string(" density");
```

Comment:
`FieldTmpOperation< FrameSolver, ParticleType >` should also have a `getName`() you can use.